source/core/ut_output_buffer.pkb

```diff
@@ -0,0 +1,100 @@
+create or replace package body ut_output_buffer is
+
+  procedure send_line(a_reporter ut_reporter_base, a_text varchar2) is
+    l_text_list  ut_varchar2_list;
+    pragma autonomous_transaction;
+  begin
+    if a_reporter is not null and a_reporter.reporter_id is not null and a_reporter.start_date is not null and a_text is not null then
+      if length(a_text) > ut_utils.gc_max_storage_varchar2_len then
+        l_text_list := ut_utils.clob_to_table(a_text, ut_utils.gc_max_storage_varchar2_len);
+        forall i in 1 .. l_text_list.count
+          insert into ut_output_buffer_tmp(start_date, reporter_id, message_id, text)
+          values (a_reporter.start_date, a_reporter.reporter_id, ut_message_id_seq.nextval, l_text_list(i));
+      else
+        insert into ut_output_buffer_tmp(start_date, reporter_id, message_id, text)
+        values (a_reporter.start_date, a_reporter.reporter_id, ut_message_id_seq.nextval, a_text);
+      end if;
+      commit;
+    end if;
+  end;
+
+  procedure close(a_reporter ut_reporter_base) is
+    pragma autonomous_transaction;
+  begin
+    insert into ut_output_buffer_tmp(start_date, reporter_id, message_id, is_finished)
+    values (a_reporter.start_date, a_reporter.reporter_id, ut_message_id_seq.nextval, 1);
+    commit;
+  end;
+
+  function get_lines(a_reporter_id varchar2, a_timeout_sec naturaln := gc_max_wait_sec) return ut_varchar2_list pipelined is
+    pragma autonomous_transaction;
+    l_results        ut_varchar2_list;
+    l_wait_wait_time number(10,1) := 0;
+    c_sleep_time     number(1,1)  := 0.1; --sleep for 100 ms between checks
+    l_finished       boolean := false;
+  begin
+    loop
+      delete from (
+        select *
+          from ut_output_buffer_tmp where reporter_id = a_reporter_id order by message_id
+        )
+      returning text bulk collect into l_results;
+
+      --nothing fetched from output, wait and try again
+      if l_results.count = 0 then
+        dbms_lock.sleep(c_sleep_time);
```

Comment:
**I'm having a second thought.**
Maybe we should use `dbms_pipe` for signalling between sessions.
That would be optional, and if users don't want to use it, they get the option with `dbms_lock`

**How signalling with `dbms_pipe` would work?**
We would create one pipe for one reporter instance.
After saving data to buffer reporter sends signal, that new data has arrived.
Consumer (`read_lines`), first waits on a pipe_message, and when received, reads (and deletes) all data from buffer.
Messages would be very lightweight - it is just a ping to wakeup the process.

So the clients installing utPLSQL would need to decide. Either they use `dbms_pipe` and have lightweight signalling (push notification to consumer).
Or they don't want to use it, and then the `read_lines` uses a pull mechanism using `dbms_lock` to minimise the frequency of SQL statements being issued.

The install could have option to choose (two separate install scripts?).
We would use conditional compilation in `ut_output_buffer.get_lines` to decide if `dbms_pipe` or `dbms_lock` is used.